ts/Series/HeikinAshi/HeikinAshiSeries.ts

```diff
@@ -0,0 +1,364 @@
+ *
+ * @augments Highcharts.Series
+ */
+class HeikinAshiSeries extends CandlestickSeries {
+
+    /* *
+     *
+     *  Static Properties
+     *
+     * */
+
+    /**
+     * An HeikinAshi series is a style of financial chart used to describe price
+     * movements over time. It displays open, high, low and close values per
+     * data point.
+     *
+     * @sample stock/demo/heikinashi/
+     *         Heikin Ashi series
+     *
+     * @extends      plotOptions.candlestick
+     * @product      highstock
+     * @requires  modules/heikinashi
+     * @optionparent plotOptions.heikinashi
+     */
+    public static defaultOptions: HeikinAshiSeriesOptions = merge(CandlestickSeries.defaultOptions, {
+
+        dataGrouping: {
+            groupAll: true
+        }
+
+    } as HeikinAshiSeriesOptions);
+
+    /* *
+     *
+     *  Properties
+     *
+     * */
+
+    public data: Array<HeikinAshiPoint> = void 0 as any;
+
+    public heikiashiData: Array<Array<number>> = [];
+
+    public options: HeikinAshiSeriesOptions = void 0 as any;
+
+    public points: Array<HeikinAshiPoint> = void 0 as any;
+
+    public yData: Array<Array<number>> = void 0 as any;
+
+    public processedYData: Array<Array<(number|null)>> = void 0 as any;
+
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    /* eslint-disable valid-jsdoc */
+
+    /**
+     * Assign haikinashi data into the points.
+     * @private
+     *
+     * @function Highcharts.seriesTypes.heikinashi#translatePoints
+     *
+     * @return {void}
+     *
+     */
+    public assignDataPoints(): void {
+        const series = this,
+            points = series.points,
+            heikiashiData = series.heikiashiData;
+        let cropStart = series.cropStart || 0;
+
+        // Rested the proccesed data.
+        series.processedYData.length = 0;
+
+        // Modify points.
+        for (let i = 0; i < points.length; i++) {
+            const point = points[i],
+                heikiashiDataPoint = heikiashiData[i + cropStart];
+
+            point.open = heikiashiDataPoint[0];
+            point.high = heikiashiDataPoint[1];
+            point.low = heikiashiDataPoint[2];
+            point.close = heikiashiDataPoint[3];
+
+            const dataPoint = [point.open, point.high, point.low, point.close];
+
+            series.processedYData.push(dataPoint);
+        }
+    }
+
+    /**
+     * Calculate data set for the heikinashi series before creating the points.
+     * @private
+     *
+     * @function Highcharts.seriesTypes.heikinashi#getProcessedData
+     *
+     * @return {void}
+     *
+     */
+    public getHeikinashiData(): void {
+        const series = this,
+            processedYData = series.allGroupedData || series.yData, // procesed and grouped data
+            heikiashiData = series.heikiashiData;
+
+        if (!heikiashiData.length && processedYData && processedYData.length) {
+            const firstPoint: any = processedYData[0];
```

Comment:
Why casting to `any`? The same below:
```
                const dataPoint: any = processedYData[i],
                    previousDataPoint: any = heikiashiData[i - 1];
```

Reply:
The `firstPoint` is an element from the `processedYData` so it might have a type like: `number|null|undefined`.
However, after removing the `undefined` from all definitions everything works as expected. 
(I guess this type might be omitted but I didn't want to make a mess in that PR.)

So my only concern was the `number|null` types. 
In calculation, adding `null` to something doesn't change the result, so I cast it to any to avoid checks or conversion from `null` to `0`.
Is there some easy hack to avoid that casting?

Reply:
Comment about that is added below because previously I cast while invoking the method.

`this.modifyDataPoint(dataPoint, previousDataPoint);`

Reply:
Okay, thanks for the explanation!